cmd/snap-confine/snap-confine.c

```diff
@@ -390,8 +390,34 @@ int main(int argc, char **argv)
 	}
 	// TODO: check for similar situation and linux capabilities.
 	if (geteuid() == 0) {
+		/* snap-confine uses privately-shared /run/snapd/ns to store bind-mounted
+		 * mount namespaces of each snap. In the case that snap-confine is invoked
+		 * from the mount namespace it typically constructs, the said directory
+		 * does not contain mount entries for preserved namespaces as those are
+		 * only visible in the main, outer namespace.
+		 *
+		 * In order to operate in such an environment snap-confine must first
+		 * re-associate its own process with another namespace in which the
+		 * /run/snapd/ns directory is visible. The most obvious candidate is pid
+		 * one, which definitely doesn't run in a snap-specific namespace, has a
+		 * predictable PID and is long lived.
+		 */
+		sc_reassociate_with_pid1_mount_ns();
+		// Do global initialization:
+		int global_lock_fd = sc_lock_global();
+		// Ensure that "/" or "/snap" is mounted with the
+		// "shared" option on legacy systems, see LP:#1668659
+		debug("ensuring that snap mount directory is shared");
+		sc_ensure_shared_snap_mount();
+		unsigned int experimental_features = 0;
+		if (sc_feature_enabled(SC_FEATURE_PARALLEL_INSTANCES)) {
+			experimental_features |= SC_FEATURE_PARALLEL_INSTANCES;
+		}
+		sc_initialize_mount_ns(experimental_features);
+		sc_unlock(global_lock_fd);
```

Comment:
IME, this is changing the behavior of all classic snaps, not just parallel installed or when the experimental feature is enabled. Ie, we're reassociating, running sc_ensure_shared_snap_mount() and running parts of sc_initialize_mount_ns(). The description of the PR indicated (to me) that we would only be changing behavior conditional on parallel installs (with the understanding that the experimental feature would need to be enabled). Should the PR description be updated?

Reply:
Yes, I think it should.

To be clear, the change was driven by my desire to understand the impact. We may decide that it is too disruptive and pull back but initially I want to see where this leads to.

Reply:
I've pulled out this piece into a separate PR #7395 where some regression potential was discussed as well.

Reply:
Thank you for the the changes in https://github.com/snapcore/snapd/pull/7302/commits/77bf07d8630ac9edecefca468ecefa952a8a2abd which addresses my concern here.